### PR TITLE
Add table constructor expression parsing

### LINKED ISSUE #23021: Add table constructor expression parsing
**Description:**
Refer: https://ballerina.io/ballerina-spec/spec.html#table-constructor-expr

**Suggested Labels:**
Component/Parser, Type/NewFeature

### PR DESCRIPTION
**Pull request by lochana-chathura**
Fixes #23021 

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 1]
```diff
@@ -7061,4 +7070,171 @@ private STNode parseTypeCastParam() {
...
+        startContext(ParserRuleContext.TABLE_CONSTRUCTOR);
+        STNode tableKeyword = parseTableKeyword();
+        STToken nextToken = peek();
+
+        STNode keySpecifier;
+        if (nextToken.kind == SyntaxKind.KEY_KEYWORD) {
+            keySpecifier = parseKeySpecifier();
+        } else {
+            keySpecifier = STNodeFactory.createEmptyNode();
+        }
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:7093]: Will this have a problem if there is some additional token before `key` keyword:
for eg: `table foo key() [{k1:v1}]`.

Comment (SupunS) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:7093]: I think we'll have to do something like:
```
switch(nextTokenKind)
    case KEY: 
        // parse key-specifier
    case OPEN_BRACKET: 
        // parse row-list 
    default: 
        // if non of the above, then recover and re-parse/return
```

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 2]
```diff
@@ -272,6 +272,21 @@
     private static final ParserRuleContext[] TYPE_CAST_PARAM_RHS =
             { ParserRuleContext.TYPE_DESCRIPTOR, ParserRuleContext.GT };
 
+    private static final ParserRuleContext[] TABLE_KEYWORD_RHS =
+            { ParserRuleContext.KEY_SPECIFIER, ParserRuleContext.OPEN_BRACKET };
+
+    private static final ParserRuleContext[] TABLE_CONSTRUCTOR_RHS =
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:278]: Would it be better to call this as row-list-rhs?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 3]
```diff
@@ -272,6 +272,21 @@
...
+            { ParserRuleContext.CLOSE_BRACKET, ParserRuleContext.MAPPING_CONSTRUCTOR };
+
+    private static final ParserRuleContext[] TABLE_CONSTRUCTOR_END =
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:281]: how about table-row-end?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java [hunk 4]
```diff
@@ -272,6 +272,21 @@
...
+            { ParserRuleContext.COMMA, ParserRuleContext.CLOSE_BRACKET };
+
+    private static final ParserRuleContext[] KEY_SPECIFIER_RHS =
+            { ParserRuleContext.CLOSE_PARENTHESIS, ParserRuleContext.VARIABLE_NAME };
+
+    private static final ParserRuleContext[] KEY_SPECIFIER_END =
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParserErrorHandler.java:287]: table-key-rhs?

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java (hunk 1 above)
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java]: fixed and two cases added.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java [hunk 5]
```diff
@@ -7061,4 +7072,196 @@ private STNode parseTypeCastParam() {
...
+        STNode rowList;
+        STNode closeBracket;
+
+        // Parse optional key specifier
+        if (nextTokenKind == SyntaxKind.KEY_KEYWORD) {
+            keySpecifier = parseKeySpecifier();
+            nextTokenKind = peek().kind;
+        }
+
+        switch (nextTokenKind) {
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:7107]: I sense an issue here :) 

Try something like: `table key(a) foo key(b) []`

Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/BallerinaParser.java:7101]: Fixed and a test cases added.

**Pull request merged**